QUARKUS-1079.md

```diff
@@ -0,0 +1,74 @@
+# QUARKUS-1079 - Hibernate Reactive as Tech Preview
+
+JIRA link: https://issues.redhat.com/browse/QUARKUS-1079
+
+For native mode similar test coverage as for JVM mode needs to be ensured.
+
+## Scope of the testing
+This feature added previews of two extensions: 
+ - Hibernate Reactive
+ - Hibernate Reactive with Panache
+
+Following test coverage should be implemented for Hibernate Reactive:
+- Usage of application.properties file for passing reactive-hibernate settings [examples]
+- Usage of import.sql [examples]
+- Smoke test(getting a working Mutiny session) [examples]
+- Retrieving Multi result from the DB [examples]
+- Retrieving Uni result(require a transformation) from the DB
+- Usage of DB-generated ids and ids generated with Java code(via ReactiveIdentifierGenerator) [examples]
+- Validation of values [examples, doesn't work as expected]
+- Attribute converters
+- Usage of CompletionStages alongside a Mutiny API
+- Usage of session opened through openSession() method as well as withSession() [examples]
+- Usage of transaction started through openSession() method as well as withSession()
+- Usage of session and transaction opened with the same method
+- Session methods: find[examples], persist[examples], remove, merge, flush[examples]
+- Queries: regular[examples], native, named, JPA criteria
+- Query methods: setParameter[examples], setMaxResults, getFirstResult, getSingleResult[examples], getResultList[examples], executeUpdate
+- Eager[examples] and lazy fetches
+- Field-level lazy fetching(Hibernate team "doesn't encourage" using this feature, though)
+- Stateless sessions and their methods
+ 
+Following test coverage should be implemented for Hibernate Reactive with Panache:
+- Active record pattern
+- Repository pattern
+- list and stream methods
+- find method
+- sorting 
+- Named queries
+- DTO projection
+- Transactions
+- Id generation
+
+Ideally, all of aforementioned tests should run for PostgreSQL[examples], MySQL and DB2 connections.
```

Comment:
There are quite a lot of items listed above.

Do you have estimate how long it would take to create such coverage?

Did you check what's covered under https://github.com/quarkusio/quarkus/tree/main/integration-tests ?

Reply:
Most of them are trivial, so I would say between two or three human-weeks.

Now I did. Which tests should go into integration tests and which into our new test suite? As I can see, our ":new" TS supports OpenShift deployment and it's easier to support new databases in there. Are there any other benefits?

Reply:
quarkus/tree/main/integration-tests ... the main idea is that we should not duplicate test coverage for existing coverage. I was assuming that some parts of your proposal can be already implemented and executed in upstream as part of every PR in both JVM and Native mode.

Areas that are not covered are fine to be in our new TS.

Reply:
As far as I can see, most(all?) areas are covered in integration tests. The thing is, there is a lack of coverage in terms of databases(panache is tested only with postgres, there is no coverage for oracle and mssql at all). Also, as far as I understand, they do not test anything on OpenShift(or Kubernetes).

So, I suggest we put these thing into our TS:
1) Basic set-up operations(app.properties, sql-load-script)
2) Basic operations for every database in(find, persist, remove, merge,flush)
3) Advanced things, if we suspect, that they may differ between databases(queries, id generation, transactions)
All of aforementioned should be implemented either for pure reactive Hibernate or for Panache, depends in which one is more popular among RH clients.
On the other hand, we should implement these things in integration tests:
1) Attribute converters
2) Tests currently implemented for Hibernate, if we will implement only Panache tests or vice versa.

What do you think?

Reply:
> there is a lack of coverage in terms of databases(panache is tested only with postgres, there is no coverage for oracle and mssql at all).

imho we should contribute upstream to enhance the set, at least for mysql/mariadb and mssql.

> they do not test anything on OpenShift(or Kubernetes).

That's true, on OpenShift we do not need to cover every scenario tested in integration-tests
Upstream is working on introducing OpenShift an Kubernetes tests as part of the CI.

About our TS, I would start with pure Reactive Hibernate, reactive support for Panache is quite new.

Reply:
The biggest problem of integration-tests — it's a zoo. There are four maven modules(postgres, mysql, db2, panache) and each one contains different methods and entity classes, which duplicate each other. In order to check a behaviour for different db's, we should create new test at least three times and each time it should work with a different data class.
As alternative, we can create a separate "unified" module, but I'm afraid it will only worsen the situation[1]

It's interesting, since Reactive Panache looks more polished(I found that field validation works in Panache HR, but doesn't work in Standard HR) and guide for it is at least in the works[2](no Hibernate Reactive guide exists at the moment).

[1] https://xkcd.com/927/
[2] https://github.com/quarkusio/quarkus/issues/18103

Reply:
so this sounds like we skip integration-testsuite for now and focus just on test development in our TS
about docs ... ping Sanne on the issue about the non-panache guide and plans about it
we can revisit integration-tests situation once Hibernate Reactive moves towards supported level

Please adjust the TP to say that we focus on test development in our repo